addon/Feeds/TopSitesFeed.js

```diff
@@ -10,26 +10,60 @@ const getScreenshot = require("addon/lib/getScreenshot");
 const {isRootDomain} = require("addon/lib/utils");
 
 Cu.import("resource://gre/modules/Task.jsm");
+Cu.import("resource://gre/modules/NewTabUtils.jsm");
 
 module.exports = class TopSitesFeed extends Feed {
-  constructor(...args) {
-    super(...args);
+  constructor(options) {
+    super(options);
     this.getScreenshot = getScreenshot;
     this.missingData = false;
+    this.pinnedLinks = options.pinnedLinks || NewTabUtils.pinnedLinks;
   }
   shouldGetScreenshot(link) {
     const isMissingIcon = !link.favicon_width && !link.favicon_height;
     const hasSmallIcon = !isMissingIcon && ((link.favicon_width < MIN_ICON_SIZE) || (link.favicon_height < MIN_ICON_SIZE));
     const badIcon = link.hasMetadata && (isMissingIcon || hasSmallIcon);
     return !isRootDomain(link.url) || badIcon;
   }
+  pinLink(action) {
+    this.pinnedLinks.pin(action.data.site, action.data.index);
+  }
+  unpinLink(action) {
+    this.pinnedLinks.unpin(action.data.site);
+  }
+  sortLinks(links, pinned) {
+    // Seperate out the pinned from the rest
+    let pinnedLinks = links.filter(link => this.pinnedLinks.isPinned(link));
```

Comment:
nit: looks like the `links` was formed as "[pinnedLinks, nonPinnedFrecentLinks]". Can we just slice it into two sub-arrays instead of filtering twice here?